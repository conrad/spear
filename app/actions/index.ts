export * from './searchForm';
export * from './searches';
export * from "./results";
