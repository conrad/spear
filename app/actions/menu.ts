import { actionCreatorVoid } from './helpers';

export const toggleMenu = actionCreatorVoid('TOGGLE_MENU');

export function toggleMenuDropDown(): Function {
  return (dispatch: Function) => {
    dispatch(toggleMenu());
  }
}
