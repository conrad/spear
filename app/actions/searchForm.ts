import { actionCreatorVoid, actionCreator } from './helpers';

export const submitSearch = actionCreatorVoid('SUBMIT_SEARCH');
export const unsetPhrase = actionCreator<number>('DELETE_PHRASE');
export const resetFile = actionCreatorVoid('RESET_FILE');
export const setFile = actionCreator<string>('SET_FILE');
export const appendPhrase = actionCreator<string>('APPEND_PHRASE');
export const setAddingPhrase = actionCreator<boolean>('SET_ADDING_PHRASE');

export function addFile(filename: string) {
  return (dispatch: Function, getState: Function) => {
    // const { searchForm } = getState();
    dispatch(setFile(filename));
  }
}

export function addPhrase(phrase: string) {
  return (dispatch: Function, getState: Function) => {
    // const { searchForm } = getState();
    console.log('addphrase in actions');
    dispatch(appendPhrase(phrase));
    dispatch(setAddingPhrase(false));
  }
}

export function deletePhrase(index: number) {
  return (dispatch: Function, getState: Function) => {
    // const { searchForm } = getState();
    dispatch(unsetPhrase(index));
  }
}

export function startAddingPhrase() {
  return (dispatch: Function) => {
    dispatch(setAddingPhrase(true));
  };
}

export function submitValidSearch() {
  return (dispatch: Function, getState: Function) => {
    const { searchForm } = getState();

    if (!searchForm.filename || searchForm.phrases.length < 1) {
      return;
    }

    dispatch(submitSearch());
  };
}
