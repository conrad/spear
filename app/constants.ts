export const TOGGLE_SHOW_RESULT = 'TOGGLE_SHOW_RESULT'
export const SHOW_RESULT_OVERLAY = 'SHOW_RESULT_OVERLAY'
export const TOGGLE_SHOW_RESULTS_WINDOW = 'TOGGLE_SHOW_RESULTS_WINDOW'
export const EXPORT_RESULTS = 'EXPORT_RESULTS'
export const HIDE_RESULT_OVERLAY = 'HIDE_RESULT_OVERLAY'


export const STORE_SEARCH = 'STORE_SEARCH'
export const UNSET_PHRASE = 'UNSET_PHRASE'
export const DELETE_SEARCH = 'DELETE_SEARCH'
export const MOVE_SEARCH = 'MOVE_SEARCH'
export const RESET_FILE = 'RESET_FILE'
export const SUBMIT_SEARCH = 'SUBMIT_SEARCH'
export const SET_FILE = 'SET_FILE'
export const SET_NEW_PHRASE = 'SET_NEW_PHRASE'
export const SET_IS_PHRASE_USED = 'SET_IS_PHRASE_USED'
export const TOGGLE_SEARCH_AS_USED = 'TOGGLE_SEARCH_AS_USED'
export const STORE_NEW_SEARCH = 'STORE_NEW_SEARCH'
export const SET_ACTIVE_SEARCH = 'SET_ACTIVE_SEARCH'
export const STORE_SEARCHES_FROM_PROFILE = 'STORE_SEARCHES_FROM_PROFILE'
export const EXPORT_SEARCHES_TO_FILE = 'EXPORT_SEARCHES_TO_FILE'
export const SET_NEW_SEARCH_NAME = 'SET_NEW_SEARCH_NAME'
