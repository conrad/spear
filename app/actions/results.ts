import { actionCreator, actionCreatorVoid } from './helpers';
import { IResults } from "../reducers/results";

export const saveResults = actionCreator<IResults>('SAVE_RESULTS');
export const saveResultsToFile = actionCreatorVoid('SAVE_RESULTS_TO_FILE');

export function setResults(results: IResults) {
  return (dispatch: Function) => {
    dispatch(saveResults(results));
  };
}
