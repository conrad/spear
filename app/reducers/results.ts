import { IAction } from '../actions/helpers';
import { saveResults, saveResultsToFile } from '../actions/results';
import { clone } from '../utils/helpers';

const initialState: IResults = { hasRun: false, items: [] }; 

export interface IResults {
  hasRun: boolean,
  items: Array<IResult>,
}

export interface IResult {
  search: string,
  phrase: string,
  excerpts: Array<IExcerpt>,
};

export interface IExcerpt {
  location: string,
  index: number,
  text: string,
};

export type TState = IResults;

export default function results(state: IResults = initialState, action: IAction) {
  let newState: IResults = clone(state);

  if (saveResults.test(action)) {
    newState = action.payload;
    return newState;
  } 
  if (saveResultsToFile.test(action)) {
    console.log("saving results in reducer");
    return newState;
  } 

  return newState;
}  