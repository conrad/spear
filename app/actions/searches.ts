import { actionCreator, actionCreatorVoid } from './helpers';
import { IPhrase, ISearch, IMove } from '../reducers/searches';
import JsonReader from '../local/import/searchLoader';
import JsonWriter from '../local/export/searchExporter';

export const storeSearch = actionCreator<ISearch>('STORE_SEARCH');
export const deleteSearch = actionCreator<ISearch>('DELETE_SEARCH');
export const moveSearch = actionCreator<IMove>('MOVE_SEARCH');
export const unsetPhrase = actionCreator<IPhrase>('UNSET_PHRASE');
export const toggleSearchAsUsed = actionCreator<number>('TOGGLE_SEARCH_AS_USED');
export const setNewSearchName = actionCreator<string>('SET_NEW_SEARCH_NAME');
export const storeNewSearch = actionCreatorVoid('STORE_NEW_SEARCH');
export const setActiveSearch = actionCreator<number>('SET_ACTIVE_SEARCH');
export const storeSearchesFromProfile = actionCreator<Array<ISearch>>('STORE_SEARCHES_FROM_PROFILE');
export const exportSearchesToFile = actionCreatorVoid('EXPORT_SEARCHES_TO_FILE') 

export function addSearch(search: ISearch) {
  return (dispatch: Function) => {
    dispatch(storeSearch(search));
  }
}

export function updateSearch(search: ISearch) {
  return (dispatch: Function) => {
    dispatch(storeSearch(search));
  }
}

export function reorderSearch(move: IMove) {
  return (dispatch: Function) => {
    dispatch(moveSearch(move));
  };
}

export function deletePhrase(phraseIndex: number, searchIndex: number) {
  return (dispatch: Function) => {
    const phrase: IPhrase = {
      index: phraseIndex,
      text: '',
      searchIndex,
    };
    
    dispatch(unsetPhrase(phrase));
  }    
}

export function setSearchAsUsed(index: number, isUsed: boolean|null) {
  return (dispatch: Function) => {
    dispatch(toggleSearchAsUsed(index));
  };
}

export function updateNewSearchName(name: string) {
  return (dispatch: Function) => {
    dispatch(setNewSearchName(name));
  };    
}

export function exportSearches() {
  return (dispatch: Function, getState: Function) => {
    const { searches } = getState();
    const jsonWriter = new JsonWriter();
    try {
      jsonWriter.saveProfile('searches_profile.json', searches);
    } catch (e) {
      console.log('Error saving searches:', e);
    }
    // dispatch(exportSearchesToFile());
  }
}

export function addSearchesFromProfile(file: File) {
  const searches: Array<ISearch> = JsonReader.retrieveSearchesFromFile(file.path);

  return (dispatch: Function) => {
    dispatch(storeSearchesFromProfile(searches));
  };    
}

export function addNewSearch() {
  return (dispatch: Function) => {
    dispatch(storeNewSearch());
  };     
}

export function selectSearch(index: number) {
  return (dispatch: Function) => {
    dispatch(setActiveSearch(index));
  };     
}

export function removeSearch(index: number, name: string) {
  return (dispatch: Function) => {
    const search = {
      index,
      name,
      phrases: [],
      isIncluded: false,
      isEditing: false
    };
    dispatch(deleteSearch(search));
  };
}
