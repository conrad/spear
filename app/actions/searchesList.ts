import { actionCreator } from './helpers';
import { ISearch, IMove } from '../reducers/searchesList';

export const storeSearch = actionCreator<ISearch>('STORE_SEARCH');
export const deleteSearch = actionCreator<ISearch>('DELETE_SEARCH');
export const moveSearch = actionCreator<IMove>('MOVE_SEARCH');

export function addSearch(index: number, name: string, description: string|null, phrases: Array<string>) {
  return (dispatch: Function) => {
    const search = {
      index,
      name,
      description,
      phrases,
      isIncluded: false,
      isEditing: false,
    };

    dispatch(storeSearch(search));
  }
}

export function reorderSearch(move: IMove) {
  return (dispatch: Function) => {
    dispatch(moveSearch(move));
  };
}

export function removeSearch(index: number, name: string) {
  return (dispatch: Function) => {
    const search = {
      index,
      name,
      description: '',
      phrases: [],
      isIncluded: false,
      isEditing: false
    };
    dispatch(deleteSearch(search));
  };
}
