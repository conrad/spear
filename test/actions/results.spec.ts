import { spy } from 'sinon'
import { emptyResultsSet } from '../fixtures/fixtures'
import * as actions from '../../app/actions/results'
import * as constants from '../../app/constants'

describe('results actions', () => {
  // it('should saveResults should create saveResults action', () => {
  //   expect(actions.saveResults(emptyResultsSet)).toMatchSnapshot()
  // })
  // it('should saveResultsInFile should create saveResultsInFile action', () => {
  //   expect(actions.saveResultsToFile()).toMatchSnapshot()
  // })

  it('should exportResults should create exportResults action', () => {
    expect(actions.exportResults().type).toEqual(constants.EXPORT_RESULTS)
  })

  it('should toggleShowResult should create toggleShowResult action', () => {
    const action = actions.toggleShowResult(1)
    expect(action.type).toEqual(constants.TOGGLE_SHOW_RESULT)
    expect(action.payload).toEqual(1)
  })

  it('should setResults should create saveResults action', () => {
    const fn = actions.setResults(emptyResultsSet)
    expect(fn).toBeInstanceOf(Function)
    const dispatch = spy()
    fn(dispatch)
    expect(dispatch.calledWith({ 
      type: actions.saveResults.type, 
      payload: emptyResultsSet
    })).toBe(true)
  })

  it('should toggleShowSearchResultRows should create toggleShowResult action', () => {
    const fn = actions.toggleShowSearchResultRows(1)
    expect(fn).toBeInstanceOf(Function)
    const dispatch = spy()
    fn(dispatch)
    expect(dispatch.calledWith({
      type: actions.toggleShowResult.type, 
      payload: 1
    })).toBe(true)
  })

  it('should showResultOverlay should create showResultOverlay action', () => {
    const fn = actions.showResultOverlay(0, 1)
    expect(fn).toBeInstanceOf(Function)
    const dispatch = spy()
    fn(dispatch)
    expect(dispatch.calledWith({
      type: actions.showOverlay.type, 
      payload: {
        resultIndex: 0,
        excerptIndex: 1,
      }
    })).toBe(true)
  })

  it('should hideResultOverlay should create hideResultOverlay action', () => {
    const fn = actions.hideResultOverlay()
    expect(fn).toBeInstanceOf(Function)
    const dispatch = spy()
    fn(dispatch)
    expect(dispatch.calledWith({
      type: actions.hideOverlay.type, 
    })).toBe(true)
  })

  it('should saveResultsToFile should create saveResultsToFile action', () => {
    const fn = actions.saveResultsToFile()
    expect(fn).toBeInstanceOf(Function)
    const dispatch = spy()
    const getState = () => ({ 
      searches: {
        searches: []
      },
      results: {
        items: []
      }
    })
    fn(dispatch, getState)
    expect(dispatch.calledWith({
      type: actions.exportResults.type, 
    })).toBe(true)
  })
})
