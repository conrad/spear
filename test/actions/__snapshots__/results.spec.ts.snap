// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions should saveResults should create saveResults action 1`] = `
Object {
  "payload": Object {
    "hasRun": true,
    "items": Array [],
  },
  "type": "SAVE_RESULTS",
}
`;

exports[`actions should saveResultsInFile should create saveResultsInFile action 1`] = `
Object {
  "type": "SAVE_RESULTS_TO_FILE",
}
`;
